Lab0 - Tour of rust (#13)
UPB-FILS-MA/upb-fils-ma.github.io: docs/lab/00.md

```rust
    let s = String::from("initial contents");
```
```rust
let data = "initial contents";
 
    let s = data.to_string();
 
    // the method also works on a literal directly:
    let s = "initial contents".to_string();
```

- ### Adding to a stream

We can expand a string using the **push_str** method to add a string slice.
```rust
let mut s = String::from("foo");
s.push_str("bar");
```
**The push** method takes **a single character** as a parameter and adds it to the string.

```rust
    let mut s = String::from("lo");
    s.push('l');
```
- ### Iteration Methods on Strings

The best way to operate on pieces of strings is to be explicit about whether you want characters or bytes. For individual Unicode scalar values, use the **chars** method .
```rust
for c in "Зд".chars() {
    println!("{c}");
}
```
### Run the program
We must access the director where the main.rs file is located and execute the command:

```
cargo run
```

## Exercises

:::tip 

If you don't have Rust installed, you can use Rust Playground [https://play.rust-lang.org/?version=stable&mode=debug&edition=2021] to solve the topics.

:::

> 1. Write a program that prints your name.

Reviewer: ```suggestion
1. Write a program that prints your name (**1p**).
```

What about something like this?

---
sidebar_position: 1
---

# 00 - Rust

We will use the [Rust](https://www.rust-lang.org/) programming language for the labs.

## Resources
1. The Rust Programming Language, Chapters [1](https://doc.rust-lang.org/book/ch01-00-getting-started.html), [2](https://doc.rust-lang.org/book/ch02-00-guessing-game-tutorial.html), [3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) and [5](https://doc.rust-lang.org/book/ch05-00-structs.html)
2. [Tour of Rust](https://tourofrust.com) step by step tutorial

## Basic programming language concepts for Rust

### Standard library

By default, Rust has a set of elements defined in the standard library that it imports into each program's

> application framework. This set is called the *prelude*, and you can see everything in it in the standard

Author: ```suggestion
application framework. This set is called the *prelude*, and you can look it up in the standard
```

    let x = 5;
    println!("The value of x is: {x}");
    x = 6;
    println!("The value of x is: {x}");
}
```

In this case, we will get a compilation error because we are trying to modify the value of `x` from `5` to `6`, but
`x` is immutable, so we cannot make this modification.

Although variables are immutable by default, you can make them **mutable** by adding `mut` in front of the
variable name. Adding `mut` also conveys intent to future readers of the code by indicating that other parts
of the code will modify the value of this variable.

```rust
fn main() {
    let mut x = 5;
    println!("The value of x is: {x}");
    x = 6;
    println!("The value of x is: {x}");
}
```

Now the value of `x` can become `6`.

### Constants

Like immutable variables, constants are values that are tied to a name and **are not allowed to change**,
but there are some differences between constants and variables.

First of all, you are not allowed to use `mut` with constants. Constants are not only immutable by default,
they are always immutable. You declare constants using the `const` keyword instead of the let keyword .

```rust
const THREE_HOURS_IN_SECONDS: u32 = 60 * 60 * 3;
```

> :::tip 

Reviewer: ```suggestion
:::info 
```

What about info for this type of tip?

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

### Data Types

> - ### Scalar types

Reviewer: ```suggestion
#### Scalar types
```

```

| Length              | Signed | Unsigned | Java Equivalent |
| :----------------: | :------: | :----: | :----: |
| 8-bit |   `i8`   | `u8` | `byte`/ `Byte`[^java_unsigned] |
| 16-bit |   `i16`   | `u16` | `short` / `Short`[^java_unsigned] |
| 32-bit |   `i32`   | `u32` | `int` / `Integer`[^java_unsigned] |
| 64-bit |   `i64`   | `u64` | `long` / `Long`[^java_unsigned] |
| 128-bit |   `i128`   | `u128` | N/A |
| arch |   `isize`   | `usize` | N/A |

**Floating Point** → Rust's floating point types are f32 and f64, which are 32-bit and 64-bit in size, respectively. The default type is **f64** because on modern CPUs it is about the same speed as f32 but is capable of more precision. All floating point types are **signed**.

```rust
fn main() {
    let x = 2.0; // f64
    let y: f32 = 3.0; // f32
}
```

**Boolean** → Booleans are one byte in size. Boolean type in Rust is specified using bool.

```rust
let t = true;
let f: bool = false; // with explicit type annotation
```

**Character** → The Rust char type is the most primitive alphabetic type in the language.

```rust
let c = 'z';
let z: char = 'ℤ'; // with explicit type annotation
let heart_eyed_cat = '😻';
```

> - ### Compound types

Reviewer: ```suggestion
#### Compound types
```

> **Tuple** → A tuple is a general way of grouping a number of values ​​with a variety of types into a single compound type. Tuples have a **fixed** length  : once declared, their size cannot increase or decrease.

Reviewer: The tuple is a structure.

:::

- ### Declarations vs. expressions

Function bodies consist of a series of instructions optionally ending with an expression.

Declarations are **statements** that perform an action and do not return a value.

Expressions evaluate to a resulting value .

:::tip 

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

- ### Functions with return values

Functions can return values ​​to the code that calls them. We don't name the return values, but we must declare their type after **an arrow (→)**. In Rust, the function's return value is synonymous with the value of **the final expression** in a function's body block. You can return earlier from a function by using the return keyword and specifying a value, but most functions implicitly return the last expression.

```rust
fn five() -> i32 {
    5
}
 
fn main() {
    let x = five();
    println!("The value of x is: {x}");// "The value of x is: 5"
}
```

:::tip 

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

### Control flow

> - ### If

Reviewer: ```suggestion
#### If-else
```

All if expressions start with the if keyword , followed by a condition. Optionally, we can also include an else expression.

```rust
fn main() {
    let number = 3;
 
    if number < 5 {
        println!("condition was true");
    } else {
        println!("condition was false");
    }
}
```

You can use multiple conditions by combining if and else in an else if expression:

```rust
fn main() {
    let number = 6;
 
    if number % 4 == 0 {
        println!("number is divisible by 4");
    } else if number % 3 == 0 {
        println!("number is divisible by 3");
    } else if number % 2 == 0 {
        println!("number is divisible by 2");
    } else {
        println!("number is not divisible by 4, 3, or 2");
    }
}
```

Because if is an expression, we can use it on **the right side** of a let statement to assign the result to a variable.

```rust
fn main() {
    let condition = true;
    let number = if condition { 5 } else { 6 };
 
    println!("The value of number is: {number}");//"The value of the number is 5"
}
```

> - ### Loop

Reviewer: ```suggestion
#### loop
```

The `loop` keyword tells Rust to run a block of code over and over forever or until you **explicitly** tell it to stop.

```rust
fn main() {
    loop {
        println!("again!");
    }
}
```
One use of a loop is to retry an operation that you know might fail, such as checking if a thread has finished its work. You may also need to pass the result of this operation out of the loop to the rest of your code. To do this, you can add the value you want to return after the break expression you use to stop the loop; this value will be returned out of the loop so you can use it:

```rust
fn main() {
    let mut counter = 0;
 
    let result = loop {
        counter += 1;
 
        if counter == 10 {
            break counter * 2;
        }
    };
 
    println!("The result is {result}");
}
```

> - ### While

Reviewer: ```suggestion
#### while
```

```rust
fn main() {
    let mut number = 3;
 
    while number != 0 {
        println!("{number}!");
 
        number -= 1;
    }
 
    println!("LIFTOFF!!!");
}
```

> - ### For

Reviewer: ```suggestion
#### for
```

```rust
fn main() {
    let a = [10, 20, 30, 40, 50];
 
    for element in a {
        println!("the value is: {element}");
    }
}
```

:::tip 

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

### Structures

Structs are similar to tuples, in that they both contain multiple related values . Like tuples, pieces of a structure can be of different types. Unlike tuples, in a structure you **will name** each piece of data so that the meaning of the values ​​is clear.

> To **define a structure, we enter the** *struct* keyword and name the entire structure. The name of a structure should describe the meaning of the data elements grouped together. Then, within curly brackets, we define the names and types of the data, which we call **fields**.

Reviewer: I would use less bold.

```rust
fn main() {
    let user1 = User {
        active: true,
        username: String::from("someusername123"),
        email: String::from("someone@example.com"),
        sign_in_count: 1,
    };
}
```

To access a certain member of the struct we use this syntax:

```rust
fn main() {
    let mut user1 = User {
        active: true,
        username: String::from("someusername123"),
        email: String::from("someone@example.com"),
        sign_in_count: 1,
    };
 
    user1.email = String::from("anotheremail@example.com")
}
```

:::tip 

Note that the entire instance must be **editable**  ; Rust **doesn't allow us** to mark only certain fields as mutable!

:::

As with any expression, we can construct a new instance of the structure as the last expression in the function body to implicitly return this new instance.

```rust
fn build_user(email: String, username: String) -> User {
    User {
        active: true,
        username: username,
        email: email,
        sign_in_count: 1,
    }
}
```

> - ### Tuple structs

Reviewer: ```suggestion
#### Tuple structs
```

Rust also supports structures that resemble tuples, called **tuple structs** . Tuple structures have the additional meaning provided by the structure name but do not have names associated with their fields; instead, they just have the field types. Tuple structures are useful when you want to name the entire tuple and make it a different type from other tuples, and when naming each field as in a regular structure would be wordy or redundant.

```rust
struct Color(i32, i32, i32);
struct Point(i32, i32, i32);
 
fn main() {
    let black = Color(0, 0, 0);
    let origin = Point(0, 0, 0);
}
```

:::tip 

> For a better understanding, please read[chapter 5](https://doc.rust-lang.org/book/ch05-00-structs.html) of the documentation.

Author: ```suggestion
For a better understanding, please read [chapter 5](https://doc.rust-lang.org/book/ch05-00-structs.html) of the documentation.
```

:::

### String
Rust has only one type of string in the base language, which is the string slice str which is usually seen in its borrowed form **&str** .

The **String** type , which is provided by the Rust standard library rather than encoded in the main language, is a scalable, mutable, and owned UTF-8 encoded string type .

> - ### Creating a new String

Reviewer: ```suggestion
#### Creating a new String
```

```rust
    let mut s = String::new();
```
This line creates a new empty string called `s`, which we can then load data into.

**We can use the String::from** function or the **to_string()** function to create a string from a string literal:
```rust
    let s = String::from("initial contents");
```
```rust
let data = "initial contents";
 
    let s = data.to_string();
 
    // the method also works on a literal directly:
    let s = "initial contents".to_string();
```

> - ### Adding to a stream

Reviewer: ```suggestion
#### Adding to a string
```

We can expand a string using the **push_str** method to add a string slice.
```rust
let mut s = String::from("foo");
s.push_str("bar");
```
**The push** method takes **a single character** as a parameter and adds it to the string.

```rust
    let mut s = String::from("lo");
    s.push('l');
```

> - ### Iteration Methods on Strings

Reviewer: ```suggestion
#### Iteration Methods on Strings
```

The best way to operate on pieces of strings is to be explicit about whether you want characters or bytes. For individual Unicode scalar values, use the **chars** method .
```rust
for c in "Зд".chars() {
    println!("{c}");
}
```
### Run the program

> We must access the director where the main.rs file is located and execute the command:

Reviewer: Not true, we have to be anywhere in the crate's folder.

> ```

Reviewer: ```suggestion
```bash
```

:::

### Data Types

#### Scalar types

A scalar type represents a single value. Rust has four main scalar types: integers, floating point numbers,
booleans, and characters.

**Integer** → Each variant can be signed or unsigned and has an explicit size.
```rust
let x: i8 = -2;
let y: u16 = 25;
```

| Length              | Signed | Unsigned | Java Equivalent |
| :----------------: | :------: | :----: | :----: |
| 8-bit |   `i8`   | `u8` | `byte`/ `Byte`[^java_unsigned] |
| 16-bit |   `i16`   | `u16` | `short` / `Short`[^java_unsigned] |
| 32-bit |   `i32`   | `u32` | `int` / `Integer`[^java_unsigned] |
| 64-bit |   `i64`   | `u64` | `long` / `Long`[^java_unsigned] |
| 128-bit |   `i128`   | `u128` | N/A |
| arch |   `isize`   | `usize` | N/A |

**Floating Point** → Rust's floating point types are `f32` and `f64`, which are 32-bit and 64-bit in size, respectively. The default type is `f64` because on modern CPUs it is about the same speed as `f32` but is capable of more precision. All floating point types are **signed**.

| Length              | Floating point | Java Equivalent|
| :----------------: | :------: | :----: | 

> | 32-bit | `f32` | `float`/ `Float` |
> | 64-bit | `f64` | `double` / `Double`|

Reviewer: ```suggestion
| 32-bit | `f32` | `float` |
| 64-bit | `f64` | `double` |
```

Floats are signed, so the primary types are similar.

```rust
fn main() {
    let x = 2.0; // f64

>     let y: f32 = 3.0; // f32

Reviewer: ```suggestion
    let y1: f32 = 3.0; // f32
    let y2 = 3.0f32; // f32
```

}
```

**Boolean** → Booleans are one byte in size. Boolean type in Rust is specified using bool.

```rust
let t = true;
let f: bool = false; // with explicit type annotation
```

**Character** → The Rust char type is the most primitive alphabetic type in the language.

```rust
let c = 'z';
let z: char = 'ℤ'; // with explicit type annotation
let heart_eyed_cat = '😻';
```

#### Compound types

**Tuple** → A tuple is a structure used for grouping a number of values ​​with a variety of types into a single compound type. Tuples have a **fixed** length  : once declared, their size cannot increase or decrease.

```rust
let tup: (i32, f64, u8) = (500, 6.4, 1);
```

**Array** → Unlike a tuple, each element in an array must have the **same type**. Unlike arrays in some other languages, Rust arrays have a **fixed** length.

```rust
let a = [1, 2, 3, 4, 5];
```

> :::tip 

Reviewer: ```suggestion
:::info 
```

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

### Functions

> We define a function in Rust by entering fn followed by a function name and a set of parentheses. Curly braces tell the compiler where the function body begins and ends.

Reviewer: ```suggestion
We define a function in Rust by entering `fn` keyword followed by a function name and a set of parentheses. Curly braces tell the compiler where the function body begins and ends.
```

```rust
fn main() {
    println!("Hello, world!");
 
    another_function();
}
 
fn another_function() {
    println!("Another function.");
}
```

> - ### Settings

Reviewer: What do you mean by settings?

> - ### Settings

Author: ```suggestion
- ### Parameters
```

> We can define functions with parameters, which are special variables that are part of a function's signature. When a function has parameters, you can provide it with **concrete values** ​​for those parameters.

Reviewer: ```suggestion
We can define functions with parameters, which are special variables that are part of a function's signature. When a function has parameters, you can provide it with *concrete values* ​​for those parameters, also called *arguments*.
```

```rust
fn main() {

>     another_function(5);

Reviewer: ```suggestion
    // the `another_function` function call has one single argument, the value 5.
    another_function(5);
```

}
 

> fn another_function(x: i32) {

Reviewer: ```suggestion
// the `another_function`function has one single parameter `x` of type `i32`
fn another_function(x: i32) {
```

    println!("The value of x is: {x}");
}
```

> :::tip 

Reviewer: ```suggestion
:::note 
```

In function signatures you must declare the type of each parameter!

:::

> - ### Declarations vs. expressions

Reviewer: ```suggestion
#### Declarations vs. expressions
```

Function bodies consist of a series of instructions optionally ending with an expression.

Declarations are **statements** that perform an action and do not return a value.

Expressions evaluate to a resulting value .

> :::tip 

Reviewer: ```suggestion
:::info 
```

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

> - ### Functions with return values

Reviewer: ```suggestion
#### Functions with return values
```

> Functions can return values ​​to the code that calls them. We don't name the return values, but we must declare their type after **an arrow (→)**. In Rust, the function's return value is synonymous with the value of **the final expression** in a function's body block. You can return earlier from a function by using the return keyword and specifying a value, but most functions implicitly return the last expression.

Reviewer: ```suggestion
Functions can return values ​​to the code that calls them. We don't name the return values, but we must declare their type after *an arrow* (`->`). In Rust, the function's return value is synonymous with the value of **the final expression** in a function's body block. You can return earlier from a function by using the `return` keyword and specifying a value, but most functions implicitly return the last expression.
```

```rust
fn five() -> i32 {
    5
}
 
fn main() {
    let x = five();
    println!("The value of x is: {x}");// "The value of x is: 5"
}
```

> :::tip 

Reviewer: ```suggestion
:::info 
```

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

### Control flow

> #### If-else

Reviewer: ```suggestion
#### if-else
```

> All if expressions start with the if keyword , followed by a condition. Optionally, we can also include an else expression.

Reviewer: ```suggestion
All `if` expressions start with the `if` keyword , followed by a condition. Optionally, we can also include an `else` expression.
```

```rust
fn main() {
    let number = 3;
 
    if number < 5 {
        println!("condition was true");
    } else {
        println!("condition was false");
    }
}
```

> You can use multiple conditions by combining if and else in an else if expression:

Reviewer: ```suggestion
You can use multiple conditions by combining `if` and `else` in an `else if` expression:
```

```rust
fn main() {
    let number = 6;
 
    if number % 4 == 0 {
        println!("number is divisible by 4");
    } else if number % 3 == 0 {
        println!("number is divisible by 3");
    } else if number % 2 == 0 {
        println!("number is divisible by 2");
    } else {
        println!("number is not divisible by 4, 3, or 2");
    }
}
```

> Because if is an expression, we can use it on **the right side** of a let statement to assign the result to a variable.

Reviewer: ```suggestion
Because `if` is an expression, we can use it on **the right side** of a `let` statement to assign the result to a variable.
```

```rust
fn main() {
    let condition = true;
    let number = if condition { 5 } else { 6 };
 
    println!("The value of number is: {number}");//"The value of the number is 5"
}
```
#### loop

The `loop` keyword tells Rust to run a block of code over and over forever or until you **explicitly** tell it to stop.

```rust
fn main() {
    loop {
        println!("again!");
    }
}
```

> One use of a loop is to retry an operation that you know might fail, such as checking if a thread has finished its work. You may also need to pass the result of this operation out of the loop to the rest of your code. To do this, you can add the value you want to return after the break expression you use to stop the loop; this value will be returned out of the loop so you can use it:

Reviewer: ```suggestion
One use of a `loop` is to retry an operation that you know might fail, such as checking if a thread has finished its work. You may also need to pass the result of this operation out of the loop to the rest of your code. To do this, you can add the value you want to return after the `break` expression you use to stop the loop; this value will be returned out of the loop so you can use it:
```

```rust
fn main() {
    let mut counter = 0;
 
    let result = loop {
        counter += 1;
 
        if counter == 10 {
            break counter * 2;
        }
    };
 
    println!("The result is {result}");
}
```

#### while
```rust
fn main() {
    let mut number = 3;
 
    while number != 0 {
        println!("{number}!");
 
        number -= 1;
    }
 
    println!("LIFTOFF!!!");
}
```

#### for
```rust
fn main() {
    let a = [10, 20, 30, 40, 50];
 
    for element in a {
        println!("the value is: {element}");
    }
}
```

> :::tip 

Reviewer: ```suggestion
:::info 
```

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

### Structures

Structs are similar to tuples, in that they both contain multiple related values . Like tuples, pieces of a structure can be of different types. Unlike tuples, in a structure you **will name** each piece of data so that the meaning of the values ​​is clear.

To define a structure, we enter the `struct` keyword and name the entire structure. The name of a structure should describe the meaning of the data elements grouped together. Then, within curly brackets, we define the names and types of the data, which we call **fields**.

```rust
struct User {
    active: bool,
    username: String,
    email: String,
    sign_in_count: u64,
}
```

> To use a structure after having defined it, we create an **instance** of this structure by specifying concrete values ​​for each of the fields. We create an instance by specifying the structure **name , then add curly braces containing key:value pairs** , where the keys are the field names and the values ​​are the data we want to store in those fields.

Reviewer: ```suggestion
To use a structure after having defined it, we create an **instance** of this structure by specifying concrete values ​​for each of the fields. We create a **stack allocated** instance by specifying the structure name , then add curly braces containing `key:value` pairs , where the keys are the field names and the values ​​are the data we want to store in those fields.
```

```rust
fn main() {
    let user1 = User {
        active: true,
        username: String::from("someusername123"),
        email: String::from("someone@example.com"),
        sign_in_count: 1,
    };
}
```

> To access a certain member of the struct we use this syntax:

Reviewer: ```suggestion
To access a certain member of the structure we use this syntax:
```

```rust
fn main() {
    let mut user1 = User {
        active: true,
        username: String::from("someusername123"),
        email: String::from("someone@example.com"),
        sign_in_count: 1,
    };
 
    user1.email = String::from("anotheremail@example.com")
}
```

> :::tip 

Reviewer: ```suggestion
:::note 
```

Note that the entire instance must be **editable**  ; Rust **doesn't allow us** to mark only certain fields as mutable!

:::

As with any expression, we can construct a new instance of the structure as the last expression in the function body to implicitly return this new instance.

```rust
fn build_user(email: String, username: String) -> User {
    User {
        active: true,
        username: username,
        email: email,
        sign_in_count: 1,
    }
}
```
#### Tuple structs

Rust also supports structures that resemble tuples, called **tuple structs** . Tuple structures have the additional meaning provided by the structure name but do not have names associated with their fields; instead, they just have the field types. Tuple structures are useful when you want to name the entire tuple and make it a different type from other tuples, and when naming each field as in a regular structure would be wordy or redundant.

```rust
struct Color(i32, i32, i32);
struct Point(i32, i32, i32);
 
fn main() {
    let black = Color(0, 0, 0);
    let origin = Point(0, 0, 0);
}
```

> :::tip 

Reviewer: ```suggestion
:::info 
```

For a better understanding, please read [chapter 5](https://doc.rust-lang.org/book/ch05-00-structs.html) of the documentation.

:::

### String

> Rust has only one type of string in the base language, which is the string slice str which is usually seen in its borrowed form **&str** .

Reviewer: ```suggestion
Rust has only one type of string in the core language, which is the string slice `str` which is usually seen in its borrowed form `&str`.
```

> The **String** type , which is provided by the Rust standard library rather than encoded in the main language, is a scalable, mutable, and owned UTF-8 encoded string type .

Reviewer: ```suggestion
The `String` type , which is provided by the Rust standard library rather than encoded in the main language, is a scalable, mutable, and owned UTF-8 encoded string type .
```

#### Creating a new String
```rust
    let mut s = String::new();
```
This line creates a new empty string called `s`, which we can then load data into.

> **We can use the String::from** function or the **to_string()** function to create a string from a string literal:

Reviewer: ```suggestion
We can use the `String::from` function or the `to_string` function to create a string from a string literal:
```

```rust
    let s = String::from("initial contents");
```
```rust
let data = "initial contents";
 
    let s = data.to_string();
 
    // the method also works on a literal directly:
    let s = "initial contents".to_string();
```

#### Adding to a string

> We can expand a string using the **push_str** method to add a string slice.

Reviewer: ```suggestion
We can expand a string using the `push_str` method to add a string slice.
```

```rust
let mut s = String::from("foo");
s.push_str("bar");
```
**The push** method takes **a single character** as a parameter and adds it to the string.

```rust
    let mut s = String::from("lo");
    s.push('l');
```
#### Iteration Methods on Strings

> The best way to operate on pieces of strings is to be explicit about whether you want characters or bytes. For individual Unicode scalar values, use the **chars** method .

Reviewer: ```suggestion
The best way to operate on pieces of strings is to be explicit about whether you want characters or bytes. For individual Unicode scalar values, use the `chars` method .
```

```rust
for c in "Зд".chars() {
    println!("{c}");
}
```
### Run the program
In order to run the program we may be anywhere in the crate's folder and execute the command:

```bash
cargo run
```

## Exercises

:::tip 

If you don't have Rust installed, you can use Rust Playground [https://play.rust-lang.org/?version=stable&mode=debug&edition=2021] to solve the topics.

:::

> - **2p** given by default

Reviewer: Do no say this.

> - **2p** given by default

Author: ```suggestion
```

1. Write a program that prints your name (**1p**).
2. Define two variables and assign them a numerical value. Shows the maximum value between the two without using a temporary variable (**1p**).
3. Write a function that checks if a number is divisible by n (**1p**).
4. Define an array of numbers and write code to display the maximum value (**1p**).
5. Define a structure called Computer that defines the brand, processor name, and memory size of a computer (**2p**).  

> a. Write an associated (static) function called new that creates an instance of the structure.   
> b. Write a method called display that prints all the information.

Reviewer: ```suggestion
  a. Write an associated (static) function called new that creates an instance of the structure.   
  b. Write a method called display that prints all the information.
```

6. Define an array with elements of type Computer. Write a program that displays a menu with the following options (**2p**):  

> a. print all computers  
>  b. print the computer with the largest amount of memory  

Reviewer: ```suggestion
  a. print all computers  
  b. print the computer with the largest amount of memory  
```

First of all, you are not allowed to use `mut` with constants. Constants are not only immutable by default,
they are always immutable. You declare constants using the `const` keyword instead of the let keyword .

```rust
const THREE_HOURS_IN_SECONDS: u32 = 60 * 60 * 3;
```

:::info 

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

### Data Types

#### Scalar types

A scalar type represents a single value. Rust has four main scalar types: integers, floating point numbers,
booleans, and characters.

**Integer** → Each variant can be signed or unsigned and has an explicit size.
```rust
let x: i8 = -2;
let y: u16 = 25;
```

| Length              | Signed | Unsigned | Java Equivalent |
| :----------------: | :------: | :----: | :----: |
| 8-bit |   `i8`   | `u8` | `byte`/ `Byte`[^java_unsigned] |
| 16-bit |   `i16`   | `u16` | `short` / `Short`[^java_unsigned] |
| 32-bit |   `i32`   | `u32` | `int` / `Integer`[^java_unsigned] |
| 64-bit |   `i64`   | `u64` | `long` / `Long`[^java_unsigned] |
| 128-bit |   `i128`   | `u128` | N/A |
| arch |   `isize`   | `usize` | N/A |

> **Floating Point** → Rust's floating point types are f32 and f64, which are 32-bit and 64-bit in size, respectively. The default type is **f64** because on modern CPUs it is about the same speed as f32 but is capable of more precision. All floating point types are **signed**.

Reviewer: ```suggestion
**Floating Point** → Rust's floating point types are `f32` and `f64`, which are 32-bit and 64-bit in size, respectively. The default type is `f64` because on modern CPUs it is about the same speed as `f32` but is capable of more precision. All floating point types are **signed**.
```

```
This line creates a new empty string called `s`, which we can then load data into.

We can use the `String::from` function or the `to_string` function to create a string from a string literal:
```rust
    let s = String::from("initial contents");
```
```rust
let data = "initial contents";
 
    let s = data.to_string();
 
    // the method also works on a literal directly:
    let s = "initial contents".to_string();
```

#### Adding to a string

We can expand a string using the `push_str` method to add a string slice.
```rust
let mut s = String::from("foo");
s.push_str("bar");
```
**The push** method takes **a single character** as a parameter and adds it to the string.

```rust
    let mut s = String::from("lo");
    s.push('l');
```
#### Iteration Methods on Strings

The best way to operate on pieces of strings is to be explicit about whether you want characters or bytes. For individual Unicode scalar values, use the `chars` method .
```rust
for c in "Зд".chars() {
    println!("{c}");
}
```
### Run the program
In order to run the program we may be anywhere in the crate's folder and execute the command:

```bash
cargo run
```

## Exercises

:::tip 

> If you don't have Rust installed, you can use Rust Playground [https://play.rust-lang.org/?version=stable&mode=debug&edition=2021] to solve the topics.

Reviewer: ```suggestion
If you don't have Rust installed, you can use [Rust Playground](https://play.rust-lang.org/) to solve the topics.
```

```

**Character** → The Rust char type is the most primitive alphabetic type in the language.

```rust
let c = 'z';
let z: char = 'ℤ'; // with explicit type annotation
let heart_eyed_cat = '😻';
```

#### Compound types

**Tuple** → A tuple is a structure used for grouping a number of values ​​with a variety of types into a single compound type. Tuples have a **fixed** length  : once declared, their size cannot increase or decrease.

```rust
let tup: (i32, f64, u8) = (500, 6.4, 1);
```

**Array** → Unlike a tuple, each element in an array must have the **same type**. Unlike arrays in some other languages, Rust arrays have a **fixed** length.

```rust
let a = [1, 2, 3, 4, 5];
```
:::info 

For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

:::

### Functions

We define a function in Rust by entering `fn` keyword followed by a function name and a set of parentheses. Curly braces tell the compiler where the function body begins and ends.

```rust
fn main() {
    println!("Hello, world!");
 
    another_function();
}
 
fn another_function() {
    println!("Another function.");
}
```

> - ### Parameters

Reviewer: ```suggestion
#### Parameters
```

```rust
    let s = String::from("initial contents");
```
```rust
let data = "initial contents";
 
    let s = data.to_string();
 
    // the method also works on a literal directly:
    let s = "initial contents".to_string();
```

#### Adding to a string

We can expand a string using the `push_str` method to add a string slice.
```rust
let mut s = String::from("foo");
s.push_str("bar");
```
**The push** method takes **a single character** as a parameter and adds it to the string.

```rust
    let mut s = String::from("lo");
    s.push('l');
```
#### Iteration Methods on Strings

The best way to operate on pieces of strings is to be explicit about whether you want characters or bytes. For individual Unicode scalar values, use the `chars` method .
```rust
for c in "Зд".chars() {
    println!("{c}");
}
```
### Run the program
In order to run the program we may be anywhere in the crate's folder and execute the command:

```bash
cargo run
```

## Exercises

:::tip 

If you don't have Rust installed, you can use [Rust Playground](https://play.rust-lang.org/) to solve the topics.

:::

> 

Reviewer: I suggest asking students to go through some chapters of tour of rust before doing these exercises.

Reviewer: @DTeodor-Alexaandru

Author: Wouldn't that make the laboratory too long? First 30min will be for organizational matters, then about 30min to go with them through the lab materials and after that they have to make the exercises. I wouldn t make the first lab that demanding:)))

Reviewer: They might not be able to make the exercises without the tour.

Author: Proposal:
 - 30min organizational matters + take a look at the lab materials
 - 30min Tour of Rust
 - the rest for the exercises.
 
 Let's not forget we need time to grade them.

Reviewer: Please write this in the lab, asking students to do some of the tour of rust tutorials.

```rust
    let s = String::from("initial contents");
```
```rust
let data = "initial contents";
 
    let s = data.to_string();
 
    // the method also works on a literal directly:
    let s = "initial contents".to_string();
```

#### Adding to a string

We can expand a string using the `push_str` method to add a string slice.
```rust
let mut s = String::from("foo");
s.push_str("bar");
```
**The push** method takes **a single character** as a parameter and adds it to the string.

```rust
    let mut s = String::from("lo");
    s.push('l');
```
#### Iteration Methods on Strings

The best way to operate on pieces of strings is to be explicit about whether you want characters or bytes. For individual Unicode scalar values, use the `chars` method .
```rust
for c in "Зд".chars() {
    println!("{c}");
}
```
### Run the program
In order to run the program we may be anywhere in the crate's folder and execute the command:

```bash
cargo run
```

## Exercises

:::tip 

If you don't have Rust installed, you can use [Rust Playground](https://play.rust-lang.org/) to solve the topics.

:::

:::info

> Before tackling the exercises, take a look and do some [Tour of Rust](https://tourofrust.com/) tutorials.

Reviewer: Please be more exact about what chapters they need to do.
